chain-tx-enclave/tx-query-next/enclave-app/src/main.rs

```diff
@@ -1,42 +1,157 @@
-/// somehow this is needed with `cargo +nightly build --target=x86_64-fortanix-unknown-sgx`
-extern crate enclave_protocol;
+mod handler;
 
-use enclave_protocol::ENCRYPTION_REQUEST_SIZE;
-use std::io::{Read, Write};
-use std::net::{TcpListener, TcpStream};
+pub use rs_libc::alloc::*;
+
+use std::{
+    convert::TryInto,
+    io::{Read, Write},
+    net::{TcpListener, TcpStream},
+    sync::{Arc, Mutex},
+};
+
+use parity_scale_codec::{Decode, Encode};
+use rustls::{ServerConfig, ServerSession, StreamOwned};
+use thread_pool::ThreadPool;
+
+use enclave_protocol::{
+    DecryptionRequest, TxQueryInitRequest, TxQueryInitResponse, ENCRYPTION_REQUEST_SIZE,
+};
+use ra_enclave::{EnclaveRaConfig, EnclaveRaContext};
+
+use self::handler::{
+    get_random_challenge, handle_decryption_request, handle_encryption_request,
+    verify_decryption_request,
+};
 
 fn main() -> std::io::Result<()> {
-    // FIXME: this doesn't do yet anything interesting due to several missing pieces
-    println!("enclave: connecting to zmq");
-    let mut zmq_stream = TcpStream::connect("zmq")?;
-    // currently, it's not possible to pass arguments
-    // FIXME: fix that or bind via Usercall extensions?
+    std::env::set_var("RUST_LOG", "debug");
+    env_logger::init();
+
+    log::info!("Connecting to ZeroMQ");
+    let zmq_stream = Arc::new(Mutex::new(TcpStream::connect("zmq")?));
+
+    // Currently, it's not possible to pass arguments to enclave code
+    // FIXME: Fix that or bind via Usercall extensions?
     // ref: https://github.com/fortanix/rust-sgx/issues/136
     // ref: https://github.com/fortanix/rust-sgx/blob/master/examples/usercall-extension-bind/runner/src/main.rs#L26
-    let listener = TcpListener::bind("0.0.0.0:3443")?;
-    println!("enclave: listening");
+    let addrs = "0.0.0.0:3443";
+    let num_threads = 4;
+    let config = EnclaveRaConfig {
+        sp_addr: "0.0.0.0:8989".to_string(),
```

Comment:
is that a valid address? It will connect to sp, not listen?

Reply:
Yes. `tx-query` will connect to sp.

Reply:
if that's "0.0.0.0", what it will connect to? Or is this to be rebound in the runner?

Reply:
This'll connect to `ra-sp-server` running on local machine.

Reply:
> If `connect_stream` returns `None`, the default implementation of `connect_stream` is used. [docs](https://docs.rs/enclave-runner/0.3.0/enclave_runner/usercalls/trait.UsercallExtension.html#method.connect_stream)